keps/sig-api-machinery/2161-apiserver-default-labels/README.md

```diff
@@ -0,0 +1,350 @@
+can be used as selector by arbitrary components.
+
+### Non-Goals
+
+* Publishing name labels on all resources
+* Exposing arbitrary fields as selectors
+
+## Proposal
+
+We propose to label of all namespaces, by default, with a reserved label 
+(i.e. “kubernetes.io/metadata.name” or some such), to allow easy selection.
+
+If this label is missing, it is added by the apiserver
+
+If the label is incorrect or changed, the apiserver fails to validate the object
+
+
+### User Stories (Optional)
+
+See https://github.com/kubernetes/enhancements/pull/2113/files for relevant user stories.
+
+### Notes/Constraints/Caveats (Optional)
+
+
+### Risks and Mitigations
+
+* Every namespace gets bigger in size. We think this is inconsequential
+* Some user may already be using that label already. This is mitigated by the 
+fact that the whole "kubernetes.io" label prefix is reserved. However, since there
+is no validation of reserved label prefixes, such namespaces will be invalidated
+post-upgrade, i.e. no further writes allowed. For this reason, we can do one alpha
+release where we log/event if we observe usage of this key that is not compatible.
+
+## Design Details
+
+This can be implemented by modifying the defaults.go and validation.go components for the namespace API, i.e., in the defaults.go file for api/core:
+
+```
+func SetDefaults_Namespace(obj *v1.Namespace) {
+  if obj.Labels == nil {
+  	obj.Labels = map[string]string{}
+  }
+  
+  if _, found := obj.Labels[“kubernetes.io/metadata.name”]; !found {
+  	obj.Labels[“kubernetes.io/metadata.name”] = obj.Name
+  }
+}
+```
+And in the validation.go, we implement the guarantee of this namespace’s value 
+being constant WRT namespace name (which is immutable).
+
+```
+func ValidateNamespace(namespace *core.Namespace) field.ErrorList {
+	// Check if namespace.Labels[“kubernetes.io/metadata.name”] == namespace.Name
+  // if not add an error
+  // ...
+  return allErrs
+```
+
+This solves not just NetworkPolicy, but all such namespace selector APIs without 
+any disruption or version skew problems.
+
+### Test Plan
+* Add unit tests that creates a namespace and checks if the namespace contains 
+the label
+* Add a test that tries to select this namespace by the label (this should 
+return also only that namespace)
+* Try to modify the reserved label and this should return an error
+* Check the WATCH operation against existing namespaces that do not have this 
+label still work properly.
+
+### Graduation Criteria
+* This feature being enabled by default at least one release.
+
+### Upgrade / Downgrade Strategy
+* Check the WATCH situation described in test plan, otherwise populate the label 
+of namespaces without it.
+* In case of downgrade, as this is just a label it will work fine. Network Policies 
+that relies on this label may fail-close, which might be acceptable.
+
+### Version Skew Strategy
+N/A
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
```

Comment:
One of the interesting things about this proposal is that it can be back-filled out-of-tree on older releases with a simple namespace-labeling controller... would it be valuable to build/provide that as well?

Reply:
@abhiraut was thinking to own this via an admission controller / web hook I believe

Reply:
a reconciling controller component to fixup namespaces created prior to existence of the admission plugin would be needed as well

Reply:
Would this need to be a controller or a one-shot job?

Reply:
need to clarify what we mean by 'on read' above, will look into this now

Reply:
it seems like if folks wanted to back fill we could easily port this into patch releases ?   After thinking about it more it feels like a controller might be alot of engineering for not much benefit

Reply:
> we could easily port this into patch releases ?

As a policy, we don't backport features to patch releases.

> After thinking about it more it feels like a controller might be alot of engineering for not much benefit

Depending on the environment, accelerating the ability to use these selectors against older releases could be very useful.

Reply:
ok, do you want that to be part of this KEP or a community effort that is complimentary ?

Reply:
That's outside the scope of this.

Reply:
ok resolving, we can discuss the controller idea in a different forum